### PR TITLE
Fix truncation / buffer sizes

### PR DESCRIPTION
**Pull request by michaelortmann**
Found by: michaelortmann
Patch by: michaelortmann
Fixes: 

One-line summary:
Fix truncation / buffer sizes

Additional description (if needed):
Found with #1028, but can/should be merged independently before #1028.

Test cases demonstrating functionality (if applicable):
it should be possible to demonstrate a truncated notice in https://github.com/eggheads/eggdrop/blob/d5eff2a548841471b5fbd95f2c54176accbad737/src/mod/irc.mod/msgcmds.c#L491 and to demonstrate that this PR will fix this. But it should be enough to know, that this PR will fix compiler warnings that would occur after #1028.

### CODE REVIEW

File: src/mod/irc.mod/msgcmds.c [hunk 1]
```diff
@@ -351,7 +351,7 @@ static int msg_who(char *nick, char *host, struct userrec *u, char *par)
   struct chanset_t *chan;
   struct flag_record fr = { FR_GLOBAL | FR_CHAN, 0, 0, 0, 0, 0 };
   memberlist *m;
-  char s[UHOSTLEN], also[512], *info;
+  char s[NICKLEN + UHOSTLEN], also[512], *info;
```
Comment (vanosg) [inline src/mod/irc.mod/msgcmds.c:354]: #define UHOSTMAX    291 + NICKMAX /* 32 (ident) + 3 (\0, !, @) + NICKMAX */
#define UHOSTLEN     UHOSTMAX + 1

Should be just UHOSTLEN?

File: src/mod/irc.mod/msgcmds.c [hunk 2]
```diff
@@ -431,7 +431,7 @@ static int msg_who(char *nick, char *host, struct userrec *u, char *par)
 
 static int msg_whois(char *nick, char *host, struct userrec *u, char *par)
 {
-  char s[UHOSTLEN], s1[81], *s2;
+  char s[UHOSTLEN + 1], s1[143], *s2, stime[14];
```
Comment (vanosg) [inline src/mod/irc.mod/msgcmds.c:434]: #define UHOSTMAX    291 + NICKMAX /* 32 (ident) + 3 (\0, !, @) + NICKMAX */
#define UHOSTLEN     UHOSTMAX + 1

Should be just UHOSTLEN?

**Review by vanosg — COMMENTED** [review]
good catches! can you double-check these values and adjust if necessary?

File: src/mod/irc.mod/msgcmds.c (hunk 1 above)
Comment (michaelortmann) [reply src/mod/irc.mod/msgcmds.c:354]: In a sane world, yes. But eggdrops usage of UHOSTLEN and friends is borked. This here is an example of something that can be read quite often within eggdrop. Later in the code there is a line like:
`snprintf(s, sizeof s, "%s!%s", m->nick, m->userhost);`
_Note: If egg_snprintf would have been sprintf, the compiler would have catched it. This is what i am trying to to with #1028.
That snprintf cant fit into UHOSTLEN.
There is no better fix currently than oversizing the destination array.
In some distant future i hope to fix the real problem, but that would requite a quite large PR and some tests to make sure its stable. 
OK, in other words, eggdrop needs to define one UHOSTLEN without NICK, and one with NICK. That is currently not the case.

File: src/mod/irc.mod/msgcmds.c (hunk 2 above)
Comment (michaelortmann) [reply src/mod/irc.mod/msgcmds.c:434]: Similar to my comment above.
Here are the compiler warnings that would occur, if snprintf instead of egg_snprintf would have been used with a destination buffer of only UHOSTLEN:
```
gcc -fPIC -g -O2 -pipe -Wall -I. -I../../.. -I../../.. -I../../../src/mod  -DHAVE_CONFIG_H -I/usr/include -g3 -DDEBUG -DDEBUG_ASSERT -DDEBUG_MEM -DDEBUG_DNS  -DMAKING_MODS -c .././irc.mod/irc.c && mv -f irc.o ../
In file included from .././irc.mod/irc.c:69:
.././irc.mod/msgcmds.c: In function ‘msg_whois’:
.././irc.mod/msgcmds.c:465:37: warning: ‘snprintf’ output may be truncated before the last format character [-Wformat-truncation=]
  465 |         snprintf(s, sizeof s, "%s!%s", par, m->userhost);
      |                                     ^
.././irc.mod/msgcmds.c:465:9: note: ‘snprintf’ output 2 or more bytes (assuming 325) into a destination of size 324
  465 |         snprintf(s, sizeof s, "%s!%s", par, m->userhost);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.././irc.mod/msgcmds.c: In function ‘msg_who’:
.././irc.mod/msgcmds.c:385:31: warning: ‘%s’ directive output may be truncated writing up to 323 bytes into a region of size between 291 and 323 [-Wformat-truncation=]
  385 |     snprintf(s, sizeof s, "%s!%s", m->nick, m->userhost);
      |                               ^~
.././irc.mod/msgcmds.c:385:5: note: ‘snprintf’ output between 2 and 357 bytes into a destination of size 324
  385 |     snprintf(s, sizeof s, "%s!%s", m->nick, m->userhost);
      |
```

**Comment by michaelortmann** [comment]
Geo: For the uhostlen mess see also #650. Many compiler warnings are currently shaddowed by using `egg_snprintf()` instead `snprintf()`. For that see also #1028. There is more to come, but i want to see this PR merged rather soon instead of making one big PR, because not all compiler warnings about buffer sizes / truncation are equal. This here is one of the more serious ones and not only cosmetic.

File: src/mod/irc.mod/msgcmds.c (hunk 1 above)
Comment (vanosg) [reply src/mod/irc.mod/msgcmds.c:354]: To summarize this in one line, m->userhost is defined as UHOSTLEN, even though UHOSTLEN is sized to allow the nick!user@host format not the user@host format we write to it. So in theory m->userhost could be longer than expected, even though we only ever write the user@host portion to it
